lib/cloud/gcp/vm.go

```diff
@@ -0,0 +1,545 @@
+	ListInstances(ctx context.Context, projectID, location string) ([]*Instance, error)
+	// GetInstance gets a GCP VM.
+	GetInstance(ctx context.Context, req *InstanceRequest) (*Instance, error)
+	// AddSSHKey adds an SSH key to a GCP VM's metadata.
+	AddSSHKey(ctx context.Context, req *SSHKeyRequest) error
+	// RemoveSSHKey removes an SSH key from a GCP VM's metadata.
+	RemoveSSHKey(ctx context.Context, req *SSHKeyRequest) error
+}
+
+// InstancesClientConfig is the client configuration for InstancesClient.
+type InstancesClientConfig struct {
+	// InstanceClient is the underlying GCP client for the instances service.
+	InstanceClient *compute.InstancesClient
+}
+
+// CheckAndSetsDefaults checks and sets defaults for InstancesClientConfig.
+func (c *InstancesClientConfig) CheckAndSetDefaults(ctx context.Context) (err error) {
+	if c.InstanceClient == nil {
+		if c.InstanceClient, err = compute.NewInstancesRESTClient(ctx); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+// Instance represents a GCP VM.
+type Instance struct {
+	// Name is the instance's name.
+	Name string
+	// Zone is the instance's zone.
+	Zone string
+	// ProjectID is the ID of the project the VM is in.
+	ProjectID string
+	// ServiceAccount is the email address of the VM's service account, if any.
+	ServiceAccount string
+	// Labels is the instance's labels.
+	Labels map[string]string
+
+	ipAddress string
+	hostKeys  []ssh.PublicKey
+	metadata  *computepb.Metadata
+}
+
+// InstanceRequest formats an instance request based on an instance.
+func (i *Instance) InstanceRequest() InstanceRequest {
+	return InstanceRequest{
+		ProjectID: i.ProjectID,
+		Zone:      i.Zone,
+		Name:      i.Name,
+	}
+}
+
+// NewInstanccesClient creates a new InstancesClient.
+func NewInstancesClient(ctx context.Context) (InstancesClient, error) {
+	var cfg InstancesClientConfig
+	client, err := NewInstancesClientWithConfig(ctx, cfg)
+	return client, trace.Wrap(err)
+}
+
+// NewInstanccesClientWithConfig creates a new InstancesClient with custom
+// config.
+func NewInstancesClientWithConfig(ctx context.Context, cfg InstancesClientConfig) (InstancesClient, error) {
+	if err := cfg.CheckAndSetDefaults(ctx); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &instancesClient{InstancesClientConfig: cfg}, nil
+}
+
+// instancesClient implements the InstancesClient interface by wrapping
+// compute.InstancesClient.
+type instancesClient struct {
+	InstancesClientConfig
+}
+
+func isExternalNAT(s string) bool {
+	return slices.Contains([]string{
+		"external-nat",
+		"External NAT",
+	}, s)
+}
+
+func toInstance(origInstance *computepb.Instance) *Instance {
+	zoneParts := strings.Split(origInstance.GetZone(), "/")
+	zone := zoneParts[len(zoneParts)-1]
+	var ipAddress string
+outer:
+	for _, netInterface := range origInstance.GetNetworkInterfaces() {
+		for _, accessConfig := range netInterface.GetAccessConfigs() {
+			if isExternalNAT(accessConfig.GetName()) {
```

Comment:
Do we only take external node IPs, or what is this for?

Reply:
Yes, this only takes the external IPs (per other comment, I've updated this to also get the internal IP).